src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java

```diff
@@ -219,24 +244,27 @@ public List<FireStoreDirectoryEntry> batchRetrieveByPath(
   public List<String> validateRefIds(
       TableServiceClient tableServiceClient, UUID datasetId, List<String> refIdArray) {
     logger.info("validateRefIds for {} file ids", refIdArray.size());
-    return ListUtils.partition(refIdArray, MAX_FILTER_CLAUSES).stream()
-        .flatMap(
-            refIdChunk -> {
-              List<TableEntity> fileRefs =
-                  TableServiceClientUtils.batchRetrieveFiles(
-                      tableServiceClient, datasetId, refIdChunk);
-              // if no files were retrieved, then every file in list is not valid
-              if (fileRefs.isEmpty()) {
-                return refIdChunk.stream();
-              }
-              // if any files were retrieved, then remove from invalid list
-              Set<String> validRefIds =
-                  fileRefs.stream()
-                      .map(e -> e.getProperty("fileId").toString())
-                      .collect(Collectors.toSet());
-              return refIdChunk.stream().filter(id -> !validRefIds.contains(id));
-            })
-        .collect(Collectors.toList());
+    List<Future<Stream<String>>> futures = new ArrayList<>();
+    for (List<String> refIdChunk : ListUtils.partition(refIdArray, MAX_FILTER_CLAUSES)) {
+      futures.add(
+          azureTableThreadpool.submit(
+              () -> {
+                List<TableEntity> fileRefs =
+                    TableServiceClientUtils.batchRetrieveFiles(
+                        tableServiceClient, datasetId, refIdChunk);
+                // if no files were retrieved, then every file in list is not valid
+                if (fileRefs.isEmpty()) {
+                  return refIdChunk.stream();
+                }
+                // if any files were retrieved, then remove from invalid list
+                Set<String> validRefIds =
+                    fileRefs.stream()
+                        .map(e -> e.getProperty("fileId").toString())
+                        .collect(Collectors.toSet());
+                return refIdChunk.stream().filter(id -> !validRefIds.contains(id));
+              }));
+    }
+    return FutureUtils.waitFor(futures).stream().flatMap(s -> s).toList();
```

Comment:
Can you explain what the `.flatMap(s -> s)` is doing?

Reply:
Totally. I added a comment but:
```// Flat map the results of the futures since each future returns a list of refIds```